ports/atmel-samd/samd21_clocks.c

```diff
@@ -0,0 +1,632 @@
+    // rather than write to it.
+    *((uint8_t*) &GCLK->GENCTRL.reg) = gclk;
+    while (GCLK->STATUS.bit.SYNCBUSY == 1) {}
+    bool enabled = GCLK->GENCTRL.bit.GENEN;
+    common_hal_mcu_enable_interrupts();
+    return enabled;
+}
+
+void disable_gclk(uint8_t gclk) {
+    while (GCLK->STATUS.bit.SYNCBUSY == 1) {}
+    GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(gclk);
+    while (GCLK->STATUS.bit.SYNCBUSY == 1) {}
+}
+
+void connect_gclk_to_peripheral(uint8_t gclk, uint8_t peripheral) {
+    GCLK->CLKCTRL.reg = GCLK_CLKCTRL_ID(peripheral) | GCLK_CLKCTRL_GEN(gclk) | GCLK_CLKCTRL_CLKEN;
+}
+
+void disconnect_gclk_from_peripheral(uint8_t gclk, uint8_t peripheral) {
+    GCLK->CLKCTRL.reg = GCLK_CLKCTRL_ID(peripheral) | GCLK_CLKCTRL_GEN(gclk);
+}
+
+void enable_clock_generator(uint8_t gclk, uint32_t source, uint16_t divisor) {
+    uint32_t divsel = 0;
+    if (gclk == 2 && divisor > 31) {
+        divsel = GCLK_GENCTRL_DIVSEL;
+        for (int i = 15; i > 4; i++) {
+            if (divisor & (1 << i)) {
+                divisor = i - 1;
+                break;
+            }
+        }
+    }
+    GCLK->GENDIV.reg = GCLK_GENDIV_ID(gclk) | GCLK_GENDIV_DIV(divisor);
+    GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(gclk) | GCLK_GENCTRL_SRC(source) | divsel | GCLK_GENCTRL_OE | GCLK_GENCTRL_GENEN;
+    while (GCLK->STATUS.bit.SYNCBUSY != 0) {}
+}
+
+void disable_clock_generator(uint8_t gclk) {
+    GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(gclk);
+    while (GCLK->STATUS.bit.SYNCBUSY != 0) {}
+}
+
+static void init_clock_source_osc8m(void) {
+    uint32_t calib = hri_sysctrl_read_OSC8M_CALIB_bf(SYSCTRL);
+    hri_sysctrl_write_OSC8M_reg(SYSCTRL,
+                                SYSCTRL_OSC8M_FRANGE(hri_sysctrl_read_OSC8M_FRANGE_bf(SYSCTRL)) |
+                                SYSCTRL_OSC8M_CALIB(calib) |
+                                SYSCTRL_OSC8M_PRESC(3) |
+                                SYSCTRL_OSC8M_ENABLE);
+    while (!hri_sysctrl_get_PCLKSR_OSC8MRDY_bit(SYSCTRL))
+        ;
+}
+
+static void init_clock_source_osc32k(void) {
+    uint32_t calib = (*((uint32_t *)FUSES_OSC32K_CAL_ADDR) & FUSES_OSC32K_CAL_Msk) >> FUSES_OSC32K_CAL_Pos;
+    hri_sysctrl_write_OSC32K_reg(SYSCTRL,
+                                 SYSCTRL_OSC32K_CALIB(calib) |
+                                 SYSCTRL_OSC32K_EN32K |
+                                 SYSCTRL_OSC32K_ENABLE);
+    while (!hri_sysctrl_get_PCLKSR_OSC32KRDY_bit(SYSCTRL))
+        ;
+}
+
+static void init_clock_source_xosc32k(void) {
+    hri_sysctrl_write_XOSC32K_reg(SYSCTRL,
+                                  SYSCTRL_XOSC32K_EN32K |
+                                  SYSCTRL_XOSC32K_XTALEN |
+                                  SYSCTRL_XOSC32K_ENABLE);
+    while (!hri_sysctrl_get_PCLKSR_XOSC32KRDY_bit(SYSCTRL))
+        ;
+}
+
+static void init_clock_source_dfll48m(void) {
+    hri_sysctrl_write_DFLLCTRL_reg(SYSCTRL, SYSCTRL_DFLLCTRL_ENABLE);
```

Comment:
I don't think using the `hri` methods provides a lot of value. Mind switching these to the CMSIS struct form instead?

For example, this should be something like: `SYSCTRL->DFLLCTRL.reg = SYSCTRL_DFFLCTRL_ENABLE;`